File: 2020/Haskell/src/Day03.hs
module Day03
  ( day03,
  )
where

import AOC
import Data.Text (Text)
import qualified Data.Text as Text
import qualified Data.Text.IO as Text

type Parsed = [String]

type Map = [String]

isTree :: Map -> Int -> Int -> Bool
isTree map x y = ((map !! y) !! x_mod_length) == '#'
  where
    map_length = length $ head map
    x_mod_length = x `mod` map_length

getTreesSlope :: Map -> Int -> Int -> Int
getTreesSlope worldMap xStep yStep = go worldMap xStep yStep 0 0 0
  where
    yLimit = length worldMap
    go worldMap xStep yStep x y acc
      | y >= yLimit = acc
      | isTree worldMap x y = nextStep (acc + 1)
      | otherwise = nextStep acc
      where
        nextStep = go worldMap xStep yStep (x + xStep) (y + yStep)

day03 =
  let day = 3
      year = 2020
      parser x = return $ Text.unpack <$> Text.lines x
      part1 p = return $ getTreesSlope p 3 1
      part2 p =
        return $
          product $
            uncurry (getTreesSlope p)
              <$> [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
   in Runner {..}

File: 2020/rust/src/day03.rs
fn parse_input(input: &str) -> Vec<&str> {
    input.lines().collect()
}

fn is_tree(map: &[&str], x: usize, y: usize) -> bool {
    let width = map[0].len();
    map[y].as_bytes()[x % width] == b'#'
}

fn get_trees_slope(input: &[&str], x_step: usize, y_step: usize) -> usize {
    let mut x = 0;
    let mut y = 0;
    let mut count = 0;
    let y_limit = input.len();
    while y < y_limit {
        if is_tree(&input, x, y) {
            count += 1;
        }
        x += x_step;
        y += y_step;
    }
    count
}

fn solve_part1(input: &[&str]) -> usize {
    get_trees_slope(input, 3, 1)
}

fn solve_part2(input: &[&str]) -> usize {
    [
        get_trees_slope(input, 1, 1),
        get_trees_slope(input, 3, 1),
        get_trees_slope(input, 5, 1),
        get_trees_slope(input, 7, 1),
        get_trees_slope(input, 1, 2),
    ]
    .iter()
    .product()
}

pub fn run(input: String) {
    let parsed_input = parse_input(&input);
    println!("Day 03, part 1: {}", solve_part1(&parsed_input));
    println!("Day 03, part 2: {}", solve_part2(&parsed_input));
}

#[cfg(test)]
mod tests {
    use super::*;
    const INPUT: &str = "..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#";

    #[test]
    fn day01_part1_example() {
        let parsed = parse_input(INPUT);
        assert_eq!(solve_part1(&parsed), 7);
    }
    #[test]
    fn day01_part2_example() {
        let parsed = parse_input(INPUT);
        assert_eq!(solve_part2(&parsed), 336);
    }
}
